[lm75b] Add LM75B temperature sensor component

Beormund opened the pull request:
# What does this implement/fix?

Add LM75B temperature sensor component

https://www.ti.com/lit/ds/symlink/lm75b.pdf

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#5320

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [x] ESP8266
- [x] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
sensor:
  - platform: lm75b
    name: "Temperature"
    update_interval: 60s
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

Beormund committed:
Add support for LM75B temperature sensor

New LM75B temperature sensor component

Beormund committed:
Merge branch 'esphome:dev' into lm75b

swoboda1337 commented on esphome/components/lm75b/lm75b.cpp:
```diff
@@ -0,0 +1,54 @@
+float LM75BComponent::get_setup_priority() const { return setup_priority::DATA; }
+
+void LM75BComponent::dump_config() {
+  ESP_LOGCONFIG(TAG, "lm75b:");
```
Should capitalize this to be consistent with everything else.

Beormund replied:
Done

swoboda1337 commented on esphome/components/lm75b/lm75b.cpp:
```diff
@@ -0,0 +1,54 @@
+}
+
+void LM75BComponent::update() {
+  this->set_timeout("read_temp", 50, [this]() {
```
Why are you setting a timeout? What is this code waiting for?

Beormund replied:
Removed set_timeout

swoboda1337 commented on esphome/components/lm75b/lm75b.cpp:
```diff
@@ -0,0 +1,54 @@
+  });
+}
+
+int16_t LM75BComponent::twos_complement_(int16_t val, uint8_t bits) {
```
Did you test negative temps? This function doesn't seem necessary as the data sheet says the result is twos complement and `raw_temperature >>= 5` will automatically extend the sign bits.

Beormund replied:
Removed twos_complement_ and more closely aligned temperature reading code with the M5Stack demo code for the LM75B for the M5StamPLC controller: 

https://github.com/m5stack/M5StamPLC/blob/1021da52c4daa1f94e226007cfa0b69c69954f49/src/utils/lm75b/lm75b.cpp#L150-L185

Beormund committed:
Removed set_timeout, capitalised lm75b

More closely aligned temperature read code with code sample from M5Stack:  https://github.com/m5stack/M5StamPLC/blob/1021da52c4daa1f94e226007cfa0b69c69954f49/src/utils/lm75b/lm75b.cpp#L150-L185

Beormund committed:
Merge branch 'dev' into lm75b

swoboda1337 commented on esphome/components/lm75b/lm75b.cpp:
```diff
@@ -0,0 +1,47 @@
+  // Sign extend negative numbers
+  if (raw_temperature & (1 << 10)) {
+    raw_temperature |= 0xFC00;
+  }
```
I don't understand why you need to sign extend this. Did you test without it?

Signed numbers when shifted automatically sign extend. Give it a try.

swoboda1337 commented on esphome/components/lm75b/lm75b.cpp:
```diff
@@ -0,0 +1,47 @@
+float LM75BComponent::get_setup_priority() const { return setup_priority::DATA; }
+
+void LM75BComponent::dump_config() {
+  ESP_LOGCONFIG(TAG, "LM75B");
```
Should have a colon at the end

swoboda1337 reviewed (changes requested):
You also need to add tests under esphome/tests/components

Beormund committed:
Merge branch 'esphome:dev' into lm75b

Beormund committed:
Added LM75B Tests

* Restored missing colon
* Removed sign extend negative number condition and re-tested
* Added compilation tests

Beormund committed:
Merge branch 'lm75b' of https://github.com/Beormund/esphome into lm75b

swoboda1337 commented on esphome/components/lm75b/lm75b.cpp:
```diff
@@ -0,0 +1,43 @@
+
+static const char *const TAG = "lm75b";
+
+void LM75BComponent::setup() { ESP_LOGCONFIG(TAG, "Setting up LM75B..."); }
```
One last thing, should remove setup as it does nothing. Remove from the header too.

Beormund replied:
Removed

Beormund committed:
Merge branch 'esphome:dev' into lm75b

Beormund committed:
Removed setup()

Beormund committed:
Merge branch 'lm75b' of https://github.com/Beormund/esphome into lm75b

Beormund committed:
Removed unnecessary get_setup_priority

Defaults to setup_priority::DATA if not specified.

swoboda1337 commented:
The docs for TMP102 got updated: https://esphome.io/components/sensor/tmp102/

Those docs say that the TMP102 component supports the LM75. Can you give that a try? Does this component do anything different?
